internal/write/service.go

```diff
@@ -196,9 +197,30 @@ func (w *Service) HandleWrite(ctx context.Context, batch *Batch) error {
 						w.retryAttempts++
 						log.Debugf("Write proc: next wait for write is %dms\n", w.retryDelay)
 					} else {
-						log.Errorf("Write error: %s\n", perror.Error())
+						logMessage := fmt.Sprintf("Write error: %s", perror.Error())
+						logHeaders := perror.HeaderToString([]string{
+							"date",
+							"trace-id",
+							"trace-sampled",
+							"X-Influxdb-Build",
+							"X-Influxdb-Request-ID",
+							"X-Influxdb-Version",
+						})
+						if len(logHeaders) > 0 {
+							logMessage += fmt.Sprintf("\nSelect Response Headers:\n%s", logHeaders)
+						}
+						log.Error(logMessage)
+					}
+					return &http2.Error{
+						StatusCode: int(perror.StatusCode),
+						Code:       perror.Code,
+						Message: fmt.Errorf(
+							"write failed (attempts %d): %w", batchToWrite.RetryAttempts, perror,
+						).Error(),
+						Err:        perror.Err,
+						RetryAfter: perror.RetryAfter,
+						Header:     perror.Header,
```

Comment:
It doesn't look good from the API perspective to return `http.Error` from the write service. And the more serious problem is that the final message looks silly: `"400 Bad Request: write failed (attempts 0): 400 Bad Request: { \"code\": \"bad request\", \"message\": \"test header\"`.

The solution could be to create a write error that would wrap the previous:
```go
package write

type Error struct {
	origin  error
	Message string
}

func NewError(err error, message string) *Error {
	return &Error{
		origin:  err,
		Message: message,
	}
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s: %s", e.Message, e.origin)
}

func (e *Error) Unwrap() error {
	return e.origin
}
```
Then this code would be:
```go
return write.NewError(perror, fmt.Sprintf("write failed (attempts %d)", batchToWrite.RetryAttempts))
```

Reply:
Agree.  When I wrote it, I was thinking there has to be a more elegant way of implementing this.  I'm still fairly new to golang and the suggested solution did not occur to me.  Implemented as suggested, and with helper methods for extracting header values.